### ISSUE TITLE
Better understanding of MILP optimization

### ISSUE THREAD

**Issue by AlessiaLiNoce** [StanfordASL/neural-network-lyapunov#446]
Hi, I'm trying to understand the starting conditions for the state in the derivative and positive MILP formulation. In the default training, I give as input only an empty vector, so which are the values used in the MILP model?  Are they random values inside the safe box [x_up,x_lo]? Or maybe a default value defined by Gurobi itself? 
![immagine](https://github.com/StanfordASL/neural-network-lyapunov/assets/73386894/e3c8ef1d-46f6-421c-bafc-09a248a3a0ee)

**Comment by hongkai-dai** [comment]
The values of the state are un-initialized by default in our code. Internally Gurobi probably initialize them to certain values. See Gurobi's documentation on [start](https://www.gurobi.com/documentation/9.1/refman/start.html#attr:Start). In mixed-integer optimization, this warm-start should not affect the final solution you get, if the optimizer obtains the global optimal, but it can affect the solver speed. So sometimes it is beneficial to initialize the variables.

Our API allows initializing the state to certain values. You could check `x_warmstart` argument in https://github.com/StanfordASL/neural-network-lyapunov/blob/32a44d4340a354e9452abeed82a8efc2926ad2b5/neural_network_lyapunov/lyapunov.py#L229 and https://github.com/StanfordASL/neural-network-lyapunov/blob/32a44d4340a354e9452abeed82a8efc2926ad2b5/neural_network_lyapunov/lyapunov.py#L584

**Comment by AlessiaLiNoce** [comment]
I see. Also another question: how is the control action related to all optimization? The controller is not defined as a MILP constraint in the general procedure. Is it because we have the same inputs both for the controller and the Lyapunov function in the closed loop system?

**Comment by AlessiaLiNoce** [comment]
Thanks for the explanation. Just a last question: what are the differences between the train() and train_lyapunov_on_samples()? They both give Lyapunov function and controller with lyapunov condition satisfied,right?

**Comment by hongkai-dai** [comment]
>  They both give Lyapunov function and controller with lyapunov condition satisfied,right?

Right.

We implemented two algorithms, see our paper https://arxiv.org/pdf/2109.14152.pdf. `train_lyapunov_on_samples()` is for Algorithm 1 in that paper, and `train()` is for Algorithm 2 in that paper.

In both algorithms, we first solve an MIP to find the worst adversarial states. In Algorithm 1, we append the adversarial states to the training data set, then apply batched gradient descent on that training data set to fix the violation. The hope is that by fixing the violation on the training data set (with finite number of states), we can reduce the violation on all states (with uncountably infinitely many states) to zero. In Algorithm 2, we find the gradient of the maximal violation of the inner MIP problem, and then just call gradient-descent; there is not a training dataset in Algorithm 2.

Empirically, we find that algorithm 1 works better for dynamical system with small dimension (like a pendulum with 2 states), while algorithm 2 works better for dynamical systems with larger dimension (like a 3D quadrotor with 12 states).
